Geo/12646 pop for k0 procedure process (#12651)
KratosMultiphysics/Kratos: applications/GeoMechanicsApplication/custom_processes/README.md

- $c-\phi$ reduction process
- [GeoExtrapolateIntegrationPointValuesToNodesProcess](#extrapolation-of-integration-values-to-nodes)
- [ResetDisplacementProcess](#reset-displacement-process)

> - $K_0$ procedure process

Reviewer: This is for just adding the link to it for easier access:
```suggestion
- [$K_0$ procedure process](#K0-procedure-process)
```

Author: Tried, but am not sure if it was successful, due to the formula in the thing.

When this process is added to the `ProjectParameters.json`, the variables specified in `list_of_variables` can be exported as nodal output (e.g. as `nodal_results` in the `GiDOutputProcess`). 

## $K0$ procedure process
For the initialization of an in-situ stress field, the $K0$ procedure derives the horizontal effective stresses from a field of vertical effective stresses.
Pre-requisite is a computed stress field with the desired normal effective stresses in the direction indicated with "K0_MAIN_DIRECTION". The normal effective stress in "K0_MAIN_DIRECTION" remains as is. Effective normal stresses in the other two directions are affected by the $K_0$ value, all shear stresses are erased.

Depending on the given input parameters, the following scheme is adapted for computation of the $K0$ value.

> $K0_{NC}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"

Reviewer: ```suggestion
$K_0_{NC}$ is gotten from either "K_0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"
```

Author: Done, but slightly different. We started using $K_0^{nc}$ for the symbol, but the string in MaterialParameters.json really is K0_NC

> $$K0_{NC} = 1.0 - \sin \phi$$

Reviewer: ```suggestion
$$K_0_{NC} = 1.0 - \sin \phi$$
```

Author: Switched to K_0^{nc}

When "OCR" and optionally "POISSON_UNLOADING_RELOADING" are supplied, the normal consolidation value $K0_NC$ is modified:

$$K0 = OCR \cdot K0_{NC} +  \frac{\nu_{ur}}{1 - \nu_{ur}} ( OCR - 1 )$$

> $$\sigma^{'}_{initial} = \begin{bmatrix} K0 \cdot \sigma^{'}_{zz} & 0 & 0 \\ 0 & K0 \cdot \sigma^{'}_{zz} & 0 \\ 0 & 0 & \sigma^{'}_{zz} \end{bmatrix}$$

Reviewer: ```suggestion
$$\sigma^{'}_{initial} = \begin{bmatrix} K_0 \cdot \sigma^{'}_{zz} & 0 & 0 \\ 0 & K_0 \cdot \sigma^{'}_{zz} & 0 \\ 0 & 0 & \sigma^{'}_{zz} \end{bmatrix}$$
```

Author: Done.

Alternaternively, when the pre-overburden pressure "POP" is specified, the initial stress tensor becomes:

> $$\sigma^{'}_{initial} = \begin{bmatrix} K0_{NC} \cdot (\sigma^{'}_{zz} + POP ) & 0 & 0 \\ 0 & K0_{NC} \cdot (\sigma^{'}_{zz} + POP) & 0 \\ 0 & 0 & \sigma^{'}_{zz} \end{bmatrix}$$

Reviewer: ```suggestion
$$\sigma^{'}_{initial} = \begin{bmatrix} K_{0_{NC}} \cdot (\sigma^{'}_{zz} + POP ) & 0 & 0 \\ 0 & K_{0_{NC}} \cdot (\sigma^{'}_{zz} + POP) & 0 \\ 0 & 0 & \sigma^{'}_{zz} \end{bmatrix}$$
```

Author: Eventually wrote K_0^{nc} to match the Plaxis manual.

> ###Note:

Reviewer: ```suggestion
### Note:
```

Author: Done.

After the stress adaptation by the $K0$ procedure, the stress state may not be in equilibrium with the present external forces anymore. Equilibrium may be reached by performing a step without applying additional load. Reaching equilibrium may then be accomplished by movement.  

### Usage

> The process is defined as follows in json (also found in some of the [integration tests](../tests/test_k0_procedure_process)):

Reviewer: Maybe it's good to mention that this process should explicitly be added to "ProjectParameters.json".

Author: Tried, please read the new text.

```json
{
  "auxilliary_process_list": [
    {
      "python_module": "apply_k0_procedure_process",
      "kratos_module": "KratosMultiphysics.GeoMechanicsApplication",
      "process_name": "ApplyK0ProcedureProcess",
      "Parameters": {
        "model_part_name": "PorousDomain.porous_computational_model_part",
        "variable_name": "CAUCHY_STRESS_TENSOR"
      }
    }
  ]
}
```

> The apply_k0_procedure_process need the following material parameter input.

Reviewer: And maybe also for this one, to be more clear:
```suggestion
The `apply_k0_procedure_process` needs the following material parameter input to be added in the "MaterialParameters.json".
```

Author: Done.

When this process is added to the `ProjectParameters.json`, the variables specified in `list_of_variables` can be exported as nodal output (e.g. as `nodal_results` in the `GiDOutputProcess`). 

> ## $K_0$ procedure process

Reviewer 2: Could we add this to the content list in line ~7?

Author: Added

For the initialization of an in-situ stress field, the $K_0$ procedure derives the horizontal effective stresses from a field of vertical effective stresses.

> Pre-requisite is a computed stress field with the desired normal effective stresses in the direction indicated with "K0_MAIN_DIRECTION". The normal effective stress in "K0_MAIN_DIRECTION" remains as is. Effective normal stresses in the other two directions are affected by the $K_0$ value, all shear stresses are erased.

Reviewer 2: The shear stresses are only erased in case of the 'standard k0 procedure' right? I think it's a good idea to mention that here as well

Author: standard k0 procedure only refers to how the vertical stress field is computed with a modified elastic constitutive tensor. After setting the horizontal stresses, all shear stresses are always erased. See line 162 of the apply_k0_procedure_process.cpp

Reviewer 2: Ahh okay, I was mixing things up, thanks for the clarification!

Depending on the given input parameters, the following scheme is adapted for computation of the $K_0$ value.

> $K_{0_{NC}}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"

Reviewer: ```suggestion
$K_0^{nc}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"
```

Author: Done.

> $$K_{0_{NC}} = 1.0 - \sin \phi$$

Reviewer: ```suggestion
$$K_0^{nc} = 1.0 - \sin \phi$$
```

Author: Done.

> When "OCR" and optionally "POISSON_UNLOADING_RELOADING" are supplied, the normal consolidation value $K_{0_{NC}}$ is modified:

Reviewer: ```suggestion
When "OCR" and optionally "POISSON_UNLOADING_RELOADING" are supplied, the normal consolidation value $K_0^{nc}$ is modified:
```

Author: Done.

> $$K_0 = OCR \cdot K_{0_{NC}} +  \frac{\nu_{ur}}{1 - \nu_{ur}} ( OCR - 1 )$$

Reviewer: ```suggestion
$$K_0 = OCR \cdot K_0^{nc} +  \frac{\nu_{ur}}{1 - \nu_{ur}} ( OCR - 1 )$$
```

Author: Done.

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_0 \cdot \sigma^{'}_{zz}} & 0 & 0 \\

Reviewer 2: Since some of these equations don't seem to render in github, is it an idea to add some pictures instead (I agree it's not ideal, but this is not that easy to read)

Author: If in the end when merging to master there is no stable rendering, we indeed will use a picture. ( on almost all tools these formula's render o.k., so it really a github thing. )

Author: Anne helped to get the pictures in.

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_0 \cdot \sigma^{'}_{zz}} & 0 & 0 \\

Reviewer: Shall we also change this one to $K_0^{nc}$ instead of K0?

Author: On line 76 it really should be the K_0 value, the adaptation by OCR and nu_{ur} of line 74 needs to be used here.

                                         0 & {K_0 \cdot \sigma^{'}_{zz}} & 0 \\
                                         0 & 0 & {\sigma^{'}_{zz}} \end{bmatrix}$$

> Alternaternively, when the pre-overburden pressure "POP" is specified, the initial stress tensor becomes:

Reviewer 2: Minor typo
```suggestion
Alternatively, when the pre-overburden pressure "POP" is specified, the initial stress tensor becomes:
```

Author: done.

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_{0_{NC}} \cdot (\sigma^{'}_{zz} + POP )} & 0 & 0 \\

Reviewer: ```suggestion
$$\sigma^{'}_{initial} = \begin{bmatrix} {K_0^{nc}} \cdot (\sigma^{'}_{zz} + POP )} & 0 & 0 \\
```

Author: Done.

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_0^{nc} \cdot (\sigma^{'}_{zz} + POP )} - \frac{\nu_{ur}}{1 - \nu_{ur}} \cdot POP & 0 & 0 \\

Reviewer 2: In my opinion we don't really need the \cdot here (normally of course used more for inner products), but I'm also fine with keeping it

Author: Agreed, and removed.

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_0^{nc} \cdot (\sigma^{'}_{zz} + POP )} - \frac{\nu_{ur}}{1 - \nu_{ur}} \cdot POP & 0 & 0 \\

Reviewer 2: Might be my ignorance, but for me it's not directly clear what $\nu_{ur}$ means

Reviewer 2: Just looked at the example material parameters and it became clear it's the poisson unloading reloading, but still I think it'd be nice to mention that here as well

Author: tried see line 73

>                                          0 & {K_{0_{NC}} \cdot (\sigma^{'}_{zz} + POP)} & 0 \\

Reviewer: ```suggestion
                                         0 & {K_0^{nc}} \cdot (\sigma^{'}_{zz} + POP)} & 0 \\
```

Author: Done.

                                         0 & 0 & {\sigma^{'}_{zz}} \end{bmatrix}$$

### Note:
After the stress adaptation by the $K_0$ procedure, the stress state may not be in equilibrium with the present external forces anymore. Equilibrium may be reached by performing a step without applying additional load. Reaching equilibrium may then be accomplished by movement.

### Usage
The process is defined as follows in json (also found in some of the [integration tests](../tests/test_k0_procedure_process)). Without the addition of this process, no adaptation of the horizontal stresses takes place.
```json
{
  "auxilliary_process_list": [
    {
      "python_module": "apply_k0_procedure_process",
      "kratos_module": "KratosMultiphysics.GeoMechanicsApplication",
      "process_name": "ApplyK0ProcedureProcess",
      "Parameters": {
        "model_part_name": "PorousDomain.porous_computational_model_part",

>         "variable_name": "CAUCHY_STRESS_TENSOR"

Reviewer 2: What about `use_standard_procedure`?

Author: added, but we need also to write a manual text for that.

Author: Wrote a manual text.

      }
    }
  ]
}
```
The "apply_k0_procedure_process" needs the following material parameter input to be added in the "MaterialParameters.json".
```json
{
  "Variables": {

>     "K0_MAIN_DIRECTION":           1,

Reviewer 2: It's probably already how K0 is working, but one might argue that the K0_MAIN_DIRECTION is not really a material parameter, but might belong more in the process parameters (like the gravity direction inputs)

Author: I would argue that the direction and the value belong together ( I would even prefer a direction vector i.s.o. an axis number ) and would like to keep as is for a while.

    "K0_NC":                       0.6,
    "UDSM_NAME"                :  "MohrCoulomb64.dll",
    "IS_FORTRAN_UDSM"          :  true,
    "NUMBER_OF_UMAT_PARAMETERS":  6,
    "INDEX_OF_UMAT_PHI_PARAMETER": 4,
    "UMAT_PARAMETERS"          :  [30000000,
                                   0.2,
                                   1000.0,
                                   30,
                                   0.0,

>                                    1000],

Reviewer 2: If I'm correct these are not specific for k0, should we only show the ones that are? And for the rest, I think it's nice to specify which are obligatory vs optional

Author: I have added the way to recognize where PHI is here, think that it should be documented in conjunction with and as alternative for the K0_NC here, Lets discus/try together how we get a satisfactory clear way.

## $K_0$ procedure process
For the initialization of an in-situ stress field, the $K_0$ procedure derives the horizontal effective stresses from a field of vertical effective stresses.
Pre-requisite is a computed stress field with the desired normal effective stresses in the direction indicated with "K0_MAIN_DIRECTION".
The normal effective stress in "K0_MAIN_DIRECTION" remains as is.
Effective normal stresses in the other two directions are affected by the $K_0$ value, all shear stresses are erased.
A specialized method for computation of the stress field with normal effective stresses is steered with "use_standard_procedure".
When set to true, the material used for the modelpart used for the $K_0$ procedure process is changed to an incremental elastic material with constitutive tensor that has only zero off-diagonal terms and zero shear terms.
When the stress computation is completed, the original constitutive law is restored.

Depending on the given input parameters, the following scheme is adapted for computation of the $K_0$ value.

> $K_{0^{nc}}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"

Reviewer: ```suggestion
$K_{0^{nc}}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS":
```

Author: Done

> $K_{0^{nc}}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"

Reviewer: ```suggestion
$K_0^{nc}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"
```

Author: Done

> $K_{0^{nc}}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"

Reviewer 3: Shouldn't this be (to have $nc$ as superscript of $K$ rather than of $0$):
```suggestion
$K_{0}^{nc}$ is gotten from either "K0_NC" the material input file or by computation from input of "INDEX_OF_UMAT_PHI_PARAMETER" and "UMAT_PARAMETERS"
```

Author: Indeed, and Marjan gave similar comment.
Done.

$$K_0^{nc} = 1.0 - \sin \phi$$

> When the overconsolidation ratio ("OCR") and optionally the unloading-reloading Poisson's ratio $\nu_{ur}$ ("POISSON_UNLOADING_RELOADING") are supplied, the normal consolidation value $K_{0_{NC}}$ is modified:

Reviewer: ```suggestion
When the overconsolidation ratio ("OCR") and optionally the unloading-reloading Poisson's ratio $\nu_{ur}$ ("POISSON_UNLOADING_RELOADING") are supplied, the normal consolidation value $K_0^{nc}$ is modified:
```

Author: Done

> When the overconsolidation ratio ("OCR") and optionally the unloading-reloading Poisson's ratio $\nu_{ur}$ ("POISSON_UNLOADING_RELOADING") are supplied, the normal consolidation value $K_{0_{NC}}$ is modified:

Reviewer 3: I think the above suggestion also applies here (and replaced one `_` by `^`):
```suggestion
When the overconsolidation ratio ("OCR") and optionally the unloading-reloading Poisson's ratio $\nu_{ur}$ ("POISSON_UNLOADING_RELOADING") are supplied, the normal consolidation value $K_{0}^{NC}$ is modified:
```

Also, is there any reason why we use here "NC" (all uppercase) and previously "nc" (all lowercase)?

I have one more question: the text mentions that the unloading-reloading Poisson's ratio is optional. What does it use when it is not supplied? After reading the code I understand it is 0.0. Perhaps we should add the default value to the documentation?

Author: We use nc in lowercase to adopt to the same convention as the Plaxis manual.
Text about the default added.

Author: Done

> $$K_0 = OCR K_0^{nc} +  \frac{\nu_{ur}}{1 - \nu_{ur}} ( OCR - 1 )$$

Reviewer: ```suggestion
$$K_0 = OCR.K_0^{nc} +  \frac{\nu_{ur}}{1 - \nu_{ur}} ( OCR - 1 )$$
```

Author: Done

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_0 \sigma^{'}_{zz}} & 0 & 0 \\

Reviewer: I suggest we put a picture of the formula to solve this issue:
![K0_1](https://github.com/user-attachments/assets/3b81f983-5d94-4c70-96a6-2c1255eb3983)

Author: Done

                                         0 & {K_0 \sigma^{'}_{zz}} & 0 \\
                                         0 & 0 & {\sigma^{'}_{zz}} \end{bmatrix}$$

Alternatively, when the pre-overburden pressure "POP" is specified, the initial stress tensor becomes:

> $$\sigma^{'}_{initial} = \begin{bmatrix} {K_0^{nc} (\sigma^{'}_{zz} + POP )} - \frac{\nu_{ur}}{1 - \nu_{ur}} POP & 0 & 0 \\

Reviewer: ![K0_2](https://github.com/user-attachments/assets/f0c9e775-7c20-4888-9b08-ac92aefea093)

Author: Done

                                         0 & {K_0^{nc} (\sigma^{'}_{zz} + POP)} - \frac{\nu_{ur}}{1 - \nu_{ur}} POP & 0 \\
                                         0 & 0 & {\sigma^{'}_{zz}} \end{bmatrix}$$

### Note:
After the stress adaptation by the $K_0$ procedure, the stress state may not be in equilibrium with the present external forces anymore. Equilibrium may be reached by performing a step without applying additional load. Reaching equilibrium may then be accomplished by movement.

### Usage

> The process is defined as follows in json (also found in some of the [integration tests](../tests/test_k0_procedure_process)). Without the addition of this process, no adaptation of the horizontal stresses takes place.

Reviewer: ```suggestion
The process is defined as follows in "ProjectParameters.json". (also found in some of the [integration tests](../tests/test_k0_procedure_process)). Without the addition of this process, no adaptation of the horizontal stresses takes place.
```

Author: Done
